### DISCUSSION TITLE
Effect of alpha parameter on confidence intervals with bootstrapping

### DISCUSSION THREAD

**Discussion by gsketefian** [dtcenter/METplus#2545] [Statistical Computation] [answered]
Hi,

I have a METviewer question regarding the role of the alpha parameter (under the "Common" tab, in the box labeled "Conf Interval Alpha") in the generation of confidence intervals (CIs) in the plot when using aggregation statistics and bootstrapping.

After reading up on bootstrapping, my impression is that changing alpha should cause the CIs in a plot of a vx metric to change.  My understanding is that alpha represents the percentile values from both the bottom and top ends of the bootstrapped (resampled) distribution to cut off.  So setting alpha = 0.05 should put the CI endpoints at 5% and 95%, respectively, while setting it to 0.25 should cause the CI bar to become shorter and its endpoints to move to the 25% and 75% percentiles of the distribution.  (I might be off by a factor of 2, i.e. maybe it's +/- alpha/2 that are used as the cutoffs; please let me know if so.)

The particular metric I'm looking at is spread-skill of 2-m temperature.  My xml plots the skill (RMSE), the spread, and the ratio spread/skill.  I considered two cases: alpha = 0.05 and alpha = 0.25.  I've attached the plots of the two cases as well as the xml I used to generate them.  What I noticed is that the CIs on the skill and spread curves did not change with alpha, but the CIs on the ratio did.  I am wondering if this is correct; it seems to be the CIs on all curves should change.

Thanks for any clarification you can provide on this.

Gerard
@JeffBeck-NOAA @michelleharrold @mkavulich @willmayfield @VanderleiVargas-NOAA 

[plot_ss_TMP2m_nostoch_stoch.xml.txt](https://github.com/dtcenter/METplus/files/14911671/plot_ss_TMP2m_nostoch_stoch.xml.txt)
![plot_ss_TMP2m_nostoch_stoch_alpha_0p05](https://github.com/dtcenter/METplus/assets/31046882/4bc0d190-bedc-4858-b5f4-236818389666)
![plot_ss_TMP2m_nostoch_stoch_alpha_0p25](https://github.com/dtcenter/METplus/assets/31046882/72dc2fb7-885d-4eec-9967-5478ad311bae)

**Comment by jprestop** [comment]
Hi @gsketefian.  Our METviewer expert, @TatianaBurek, is out of the office currently. @TaraJensen is going to take a look and see if she is able to provide assistance.  @TatianaBurek when you return if you're able to provide additional assistance, we would appreciate it.

**Reply by JohnHalleyGotway** [reply]
@gsketefian I took a look at this issue. I was able to replicate the behavior you describe. I re-created your plot with alpha values of 0.05 and 0.25. The confidence intervals for the latter should be significantly narrower than for the former. And that is the case for the two derived ratio curves, but not for the other RMSE and Spread lines.

FYI, it's easy to see the exact difference by downloading data from the `R data` tab. I created two plots, downloaded the `R data` for both, and then ran `opendiff` on my Mac laptop to see the differences line-by-line.

I did some digging into METcalcpy, whose job it is to compute statistics and confidence intervals. On [line 102 of metcalcpy/bootstrap.py](https://github.com/dtcenter/METcalcpy/blob/e5291fc4b94a2183097059d4b55d2c7685b65922/metcalcpy/bootstrap.py#L102), I see in that the `bootstrap_and_value(... alpha=0.05 ...)` function computes 95% CI's by default. on [line 751 of metcalcpy/agg_stat.py](https://github.com/dtcenter/METcalcpy/blob/e5291fc4b94a2183097059d4b55d2c7685b65922/metcalcpy/agg_stat.py#L751), I see that the default alpha CI value is being overridden by user settings (`alpha=self.params['alpha']`), as we'd expect.  However the call to the same `bootstrap_and_value()` function on [line 826](https://github.com/dtcenter/METcalcpy/blob/e5291fc4b94a2183097059d4b55d2c7685b65922/metcalcpy/agg_stat.py#L826), DOES NOT INCLUDE an `alpha=...` setting.

Having `alpha` specified in one call to `bootstrap_and_value()` and not in another may explain the variability in confidence intervals widths you've uncovered.

I'd definitely classify this as a bug and request that @bikegeek write up a bugfix issue for it and fix it in METcalcpy... and perhaps patch METviewer to include the fix as well.

@gsketefian, for the time being, I'd recommend sticking with the default alpha value of 0.05 (corresponding to a 95% confidence interval), until we can get this bug fixed.

**Comment by bikegeek** [comment, marked as answer]
There is a Github issue: 
https://github.com/dtcenter/METcalcpy/issues/371

Please direct any further questions and comments to the comments section of the issue.

**Comment by gsketefian** [comment]
@JohnHalleyGotway Thanks for looking into this and confirming that it's a bug.  @bikegeek Thanks for opening the Github issue.  I will use alpha = 0.05 for now until the bug is fixed.

**Comment by bikegeek** [comment]
Is this a 'blocker' issue for you, or can you generate your plots with the
default alpha for now?

Regards,
Minna

**Comment by gsketefian** [comment]
@bikegeek I think we're ok with the default value of 0.05 for now, but I'll tag @JeffBeck-NOAA, @michelleharrold, @mkavulich, and @willmayfield so they can chime in as well.

Do you happen to have an estimate of how long it would take to get a bug fix in?

**Reply by michelleharrold** [reply]
Agreed that for now we can use 0.05 -- this is what we typically use, but I could definitely see us wanting to explore 0.01 or the like in the future!

**Reply by JeffBeck-NOAA** [reply]
In addition to the strictest values of statistical significance (0.01, 0.05), it would be nice to have some flexibility to assess alpha=0.1 or even 0.15 for some of our shorter retrospective simulations to see how that impacts things.

**Comment by bikegeek** [comment]
The fix in METcalcpy *should* be straightforward, it's the additional work in METviewer which will take more time.  We don't have the budget for doing any further METplus development until June of this year.

**Comment by TatianaBurek** [comment]
METviewer does not require any additional work since the bug is in METcalcpy

**Comment by bikegeek** [comment]
John HG wasn't sure whether METviewer needed any changes.  Nonetheless,
with the amount of work for the upcoming beta release, we don't have the
resources to fix this before our funding situation changes.
